src/core/expr/fexpr_count_countna.cc

```diff
@@ -0,0 +1,160 @@
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include "column/const.h"
+#include "column/latent.h"
+#include "column/countna.h"
+#include "column/countna_no_args.h"
+#include "documentation.h"
+#include "expr/fexpr_reduce_unary.h"
+#include "expr/eval_context.h"
+#include "expr/workframe.h"
+#include "python/xargs.h"
+#include "stype.h"
+namespace dt {
+namespace expr {
+
+template<bool COUNTNA>
+class FExpr_CountNA : public FExpr_ReduceUnary {
+  public:
+    using FExpr_ReduceUnary::FExpr_ReduceUnary;
+
+
+    std::string name() const override {
+      return COUNTNA? "countna"
+                    : "count";
+    }
+
+    Column evaluate1(Column&& col, const Groupby& gby, bool is_grouped) const {
+      SType stype = col.stype();
+      switch (stype) {
+        case SType::VOID:
+        case SType::BOOL:
+        case SType::INT8:
+          return make<int8_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::INT16:
+          return make<int16_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::DATE32: 
+        case SType::INT32:
+          return make<int32_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::TIME64: 
+        case SType::INT64:
+          return make<int64_t>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::FLOAT32:
+          return make<float>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::FLOAT64:
+          return make<double>(std::move(col), SType::INT64, gby, is_grouped);
+        case SType::STR32:
+        case SType::STR64:
+          return make<CString>(std::move(col), SType::INT64, gby, is_grouped);                
+        default:
+          throw TypeError()
+            << "Invalid column of type `" << stype << "` in " << repr();
+      }
+    }
+
+    template <typename T>
+    Column make(Column&& col, SType stype, const Groupby& gby, bool is_grouped) const {
+      if (is_grouped) {
+        return Column(new Latent_ColumnImpl(new Count_ColumnImpl<T, COUNTNA, true>(
+          std::move(col), stype, gby
+        )));
+      } else {
+        return Column(new Latent_ColumnImpl(new Count_ColumnImpl<T, COUNTNA, false>(
+          std::move(col), stype, gby
+        )));
+      }
+    }
+};
+
+
+// gets the count of all rows - nulls are not checked
+template<bool COUNTNA>
+class FExpr_CountNA_AllRows : public FExpr_Func {
+  public:
+    FExpr_CountNA_AllRows(){}
+
+    std::string repr() const override {
+      std::string out = COUNTNA ? "countna(None)" 
+                                : "count()";
+      return out;
+    }
+
+  Workframe evaluate_n(EvalContext &ctx) const override {
+    Workframe wf(ctx);
+    Groupby gby = ctx.get_groupby();
+    Column col;
+
+    if (!gby) {
+      gby = Groupby::single_group(wf.nrows());
+    } 
```

Comment:
This block can be removed, because `gby` is never empty in the context. If there is no `by()` clause, `gby` will be filled with a single group already.

Reply:
I'm confused here ... usually we check for an empty gby here right - is there a condition here that negates this need?

Reply:
Yes, but that check was wrong. Look here: https://github.com/h2oai/datatable/blob/main/src/core/expr/eval_context.h#L56-L58

That's why I had to push these changes: https://github.com/h2oai/datatable/blob/main/src/core/expr/fexpr_reduce_unary.cc#L35-L40